improver/lapse_rate.py

```diff
@@ -396,75 +466,38 @@ def process(self, temperature_cube, orography_cube, land_sea_mask_cube):
-            fnc = SaveNeighbourhood(allbuffers=all_temp_subsections)
-            generic_filter(temperature_data, fnc.filter,
-                           size=self.nbhood_size,
-                           mode='constant', cval=np.nan)
-
-            fnc = SaveNeighbourhood(allbuffers=all_orog_subsections)
-            generic_filter(orography_data, fnc.filter,
-                           size=self.nbhood_size,
-                           mode='constant', cval=np.nan)
-
-            # height_diff_mask is True for points where the height
-            # difference between the central point and its neighbours
-            # is > max_height_diff.
-            height_diff_mask = self._create_heightdiff_mask(
-                all_orog_subsections)
-
-            # Mask points with extreme height differences as NaN.
-            all_orog_subsections = np.where(height_diff_mask, np.nan,
-                                            all_orog_subsections)
-            all_temp_subsections = np.where(height_diff_mask, np.nan,
-                                            all_temp_subsections)
-
-            # Loop through both arrays and find gradient of each subsection.
-            # The gradient indicates lapse rate - save into another array.
-            # TODO: This for loop is the bottleneck in the code and needs to
-            # be parallelised.
-            lapse_rate_array = [self._calc_lapse_rate(temp, orog)
-                                for temp, orog in zip(all_temp_subsections,
-                                                      all_orog_subsections)]
-
-            lapse_rate_array = np.array(
-                lapse_rate_array, dtype=np.float32).reshape(dataarray_shape)
-
-            # Enforces upper and lower limits on lapse rate values.
-            lapse_rate_array = np.where(lapse_rate_array < self.min_lapse_rate,
-                                        self.min_lapse_rate, lapse_rate_array)
-            lapse_rate_array = np.where(lapse_rate_array > self.max_lapse_rate,
-                                        self.max_lapse_rate, lapse_rate_array)
-
-            lapse_rate_slice.data = lapse_rate_array
-            lapse_rate_cube_list.append(lapse_rate_slice)
-
-        lapse_rate_cube = lapse_rate_cube_list.merge_cube()
-        lapse_rate_cube.rename('air_temperature_lapse_rate')
-        lapse_rate_cube.units = 'K m-1'
+        # Create list of arrays over "realization" coordinate
+        has_realization_dimension = False
+        if temperature_cube.coords("realization"):
+            enforce_coordinate_ordering(temperature_cube, "realization")
+            temp_slices = temperature_cube.data
+            if "realization" in [coord.name() for coord in
+                                 temperature_cube.coords(dim_coords=True)]:
+                has_realization_dimension = True
+        else:
+            temp_slices = [temperature_cube.data]
+
+        # Calculate lapse rate for each realization
```

Comment:
As below - you may well be right but I haven't quite got my head around it and would rather sort if out in a separate PR (follow-on subtask is open for this).  Would that be OK?

Reply:
That's ok. I think my suggestion is the right kind of approach and the right kind of logic. It's probably not worth changing now, but it might have been better if you didn't touch this loop at all and saved all the refactoring for the later PR so it's all together. Although maybe this wasn't possible because the arrays would have mismatched in size.